### PR TITLE
[snackager] Upgrade to webpack@5 and latest babel/metro

### PR DESCRIPTION
**Pull request by IjzerenHein**
# Why

Upgrade to webpack@5, newer babel dependencies and the latest `metro-react-native-babel-preset`. To better support newer compilation scenarios; and this also produces smaller bundles.

# How

- Update dependencies to `webpack@5`
- Update to latest `metro-react-native-babel-preset`
- Update to latest `@babel/core` & `@babel/generator`
- Upgrade code where needed
- Remove no longer needed webpack@4 dependencies
- Implement babel rewrite plugin for fixing `@material-ui/core` `@babel/runtime/helper/esm/...` imports
- Update test snapshots
- Investigated whether `AssetResolver` was still needed and added comments

# Test Plan

- All tests pass
- Uses cache-busting to regenerate the following bundles and verify them using a Snack example:
  - [x] Verified `react-native-webview`
  - [x] Verified `@react-native-community/datetimepicker`
  - [x] Verified `@react-navigation/native`
  - [x] Verified `@react-navigation/stack`
  - [x] Verified `expo-linear-gradient`
  - [x] Verified `@draftbit/ui`

### CODE REVIEW

File: snackager/src/__integration-tests__/__snapshots__/bundler.test.ts.snap [hunk 1]
```diff
@@ -433,12 +433,24 @@ Object {
...
+          "./assertThisInitialized.js",
+          "./iterableToArray.js",
+          "./iterableToArrayLimit.js",
+          "./nonIterableRest.js",
+          "./nonIterableSpread.js",
+          "./objectWithoutPropertiesLoose.js",
+          "./setPrototypeOf.js",
+          "./unsupportedIterableToArray.js",
+          "@babel/runtime/helpers/interopRequireDefault",
+          "@babel/runtime/helpers/typeof",
```
Comment (IjzerenHein) [inline snackager/src/__integration-tests__/__snapshots__/bundler.test.ts.snap:453]: These externals are reported for `@drafbit/ui`. This is not right, need to investigate whether this is a fluke with the externals detection or whether externals are actually new.

Comment (IjzerenHein) [reply snackager/src/__integration-tests__/__snapshots__/bundler.test.ts.snap]: Fixed by adding a custom babel plugin that rewrites incorrect `@babel/runtime/helpers/esm/...` imports in `@material-ui/core`.

File: snackager/src/bundler/AssetResolver.ts [hunk 2]
```diff
@@ -28,52 +28,62 @@ export default class AssetResolver {
...
-                path: resolved,
-                relativePath: request.relativePath && resolver.join(request.relativePath, resolved),
-                file: true,
-              })
+    resolver
+      .getHook('file')
+      .tapAsync(
+        'SnackagerAssetResolverPlugin',
+        (request: ResolveRequest, _context: any, callback: Function) => {
+          // console.log('FILE', request.path);
```
Comment (byCedric) [inline snackager/src/bundler/AssetResolver.ts:36]: I think we can remove this, right? 😄

**Review by byCedric — APPROVED** [review]
LGTM!

**Pull request merged**